### ISSUE TITLE
Master startup fails on 3.10.0 on Twisted 23.10.0 on Python 3.12

### ISSUE THREAD

**Issue by limburgher** [buildbot/buildbot#7255]
2023-12-05 13:33:24-0600 [-] Loading buildbot.tac...
2023-12-05 13:33:25-0600 [-] Loaded.
2023-12-05 13:33:25-0600 [-] twistd 23.10.0 (/usr/bin/python3 3.12.0) starting up.
2023-12-05 13:33:25-0600 [-] reactor class: twisted.internet.epollreactor.EPollReactor.
2023-12-05 13:33:25-0600 [-] Unhandled error in Deferred:
2023-12-05 13:33:25-0600 [-] Unhandled Error
	Traceback (most recent call last):
	  File "buildbot.tac", line 28, in <module>
	    m = BuildMaster(basedir, configfile, umask)
	  File "/usr/lib/python3.12/site-packages/buildbot/master.py", line 102, in __init__
	    self._services_d = self.create_child_services()
	  File "/usr/lib/python3.12/site-packages/twisted/internet/defer.py", line 2256, in unwindGenerator
	    return _cancellableInlineCallbacks(gen)
	  File "/usr/lib/python3.12/site-packages/twisted/internet/defer.py", line 2168, in _cancellableInlineCallbacks
	    _inlineCallbacks(None, gen, status, _copy_context())
	--- <exception caught here> ---
	  File "/usr/lib/python3.12/site-packages/buildbot/master.py", line 227, in startService
	    yield self._services_d
	  File "/usr/lib/python3.12/site-packages/twisted/internet/defer.py", line 2000, in _inlineCallbacks
	    result = context.run(gen.send, result)
	  File "/usr/lib/python3.12/site-packages/buildbot/master.py", line 188, in create_child_services
	    self.www = wwwservice.WWWService()
	  File "/usr/lib/python3.12/site-packages/buildbot/www/service.py", line 196, in __init__
	    self.apps = get_plugins('www', None, load_now=True)
	  File "/usr/lib/python3.12/site-packages/buildbot/plugins/db.py", line 366, in get_plugins
	    return _DB.add_namespace(namespace, interface, load_now)
	  File "/usr/lib/python3.12/site-packages/buildbot/plugins/db.py", line 316, in add_namespace
	    tempo.load()
	  File "/usr/lib/python3.12/site-packages/buildbot/plugins/db.py", line 253, in load
	    self._tree.load()
	  File "/usr/lib/python3.12/site-packages/buildbot/plugins/db.py", line 291, in __getattr__
	    raise AttributeError(str(e)) from e
	builtins.AttributeError: Unknown component name: _tree

**Comment by p12tic** [comment]
Max version of Twisted supported by Buildbot right now is 22.10.0.

**Comment by limburgher** [comment]
True, but many distributions, Fedora in this case, have 23.10.0 on stable releases, and 3.9.2 worked with it. Can I assist with updating that twisted support?

**Comment by p12tic** [comment]
The primary problem right now is that tests in parallel mode don't work. The only thing I figured out is that someone closes fd 4 on test worker process which is used by trial to communicate between test runner and test worker. Now the fun part starts - we need to somehow put ourselves in close(4) code path and get a stacktrace.

I use Debian Bookworm with Python 3.11. For testing instruction you can check out http://docs.buildbot.net/current/developer/quickstart.html#create-a-buildbot-python-environment. After that I just modified master/setup.py to remove Twisted version limit, did `pip install -U Twisted` and ran `trial -j4 buildbot.test.unit`.

**Comment by p12tic** [comment]
And I almost forgot - thanks a lot for the offer!

**Comment by limburgher** [comment]
I'm on Fedora 39 with python 3.12 and there are LOTS of failures. I lack the time to make this a primary focus but I'll see what I can do. :)

**Comment by yan12125** [comment]
Looks like there are some ongoing efforts towards support for newer Twisted at https://github.com/buildbot/buildbot/pull/7268

**Comment by gl-yziquel** [comment]
Experienced the same issue. Please put it in the docs as long as it is not resolved.

**Comment by yan12125** [comment]
Fixes for Twisted 23.10 are merged to master branch (https://github.com/buildbot/buildbot/pull/7268, https://github.com/buildbot/buildbot/pull/7304) and backported to 3.10 branch (https://github.com/buildbot/buildbot/pull/7317, https://github.com/buildbot/buildbot/pull/7320). This issue can probably be closed.

**Comment by p12tic** [comment]
Fixed in 3.10.1.
